core/tracing/hooks.go

```diff
@@ -150,9 +150,41 @@ type (
 	// beacon block root.
 	OnSystemCallEndHook = func()
 
-	// OnSystemTxEndHook is called when tracing a system transaction, which does not calculate intrinsic gas during execution.
+	// OnSystemTxFixIntrinsicGasHook is called when tracing a system transaction, which does not calculate intrinsic gas during execution.
 	// this hook will subtract intrinsic gas from the total gas used.
-	OnSystemTxEndHook = func(uint64)
+	OnSystemTxFixIntrinsicGasHook = func(uint64)
+
+	// OnSystemTxStartHook is called when a system transaction is about to be executed within the Parlia consensus
+	// engine, like before upgrading system contracts, distribution of rewards, and other "chain" related transactions.
+	//
+	// This will be called in addition to the `OnTxStart` hook so the flow of event you will receive in your tracer will
+	// look kike this:
+	//
+	// - OnSystemTxStart
+	// - OnTxStart
+	// - OnTxEnd
+	// - OnSystemTxEnd
+	//
+	// This event flow enables transactions to be traced correctly with just OnTxStart/End being set but also
+	// enables special routing of those transactions by having OnSystemTxStart/End defined and keeping a special
+	// system state to do something different when OnTxEnd is called.
+	OnSystemTxStartHook func()
```

Comment:
It's not necessary to add `OnSystemTxStartHook` or `OnSystemTxEndHook`, it can be merged into `OnTxStart` and `OnTxEnd`, and it is more flexible too.

```go
//TxStartHook 
func(vm *VMContext, tx *types.Transaction, from common.Address) {
  // check if system contract first
  if tx.To() != nil && from == vm.Coinbase && parlia.isToSystemContract(*tx.To()) && tx.GasPrice().Cmp(big.NewInt(0)) == 0 {
      //handle system tx
   }
}
```

Reply:
`parlia` wouldn't be accessible from tracer. I agree that if within the tracer when in `OnTxStart/End` there would be a way to determine if the transaction is a system one, I think it would be ok.

However, I still prefer an explicit hooks:
- If someone does not implement `OnSystemTxStartHook/End`, everything still works.
- It makes it a bit more explicit for someone to look into the tracer Hooks and understand there could be system transaction.

But those are not super strong arguments. Let me know what you think.

Reply:
Yeah, `parlia` is not easy import in other low-level modules, it's ok for me to introduce new explicit hooks.